packages/file_selector/file_selector_windows/lib/src/file_selector_dart/file_dialog_controller.dart

```diff
@@ -0,0 +1,92 @@
+///
+/// Since this class defines the end of what can be unit tested, it should
+/// contain as little logic as possible.
+class FileDialogController {
+  /// Creates a controller managing [IFileDialog](https://pub.dev/documentation/win32/latest/winrt/IFileDialog-class.html).
+  FileDialogController(
+      IFileDialog fileDialog, IFileOpenDialogFactory iFileOpenDialogFactory)
+      : _fileDialog = fileDialog,
+        _iFileOpenDialogFactory = iFileOpenDialogFactory;
+
+  /// A reference to the file dialog.
+  final IFileDialog _fileDialog;
+
+  final IFileOpenDialogFactory _iFileOpenDialogFactory;
+
+  /// Disallow copy and assign.
+  // TODO(eugerossetto): There is no built-in clone method in Dart. In addition to this, as the default constructor has been re-writed, there is no way of passing a FileDialogController as an input parameter.
+
+  /// Sets the default folder for the dialog to [path].
+  int setFolder(Pointer<COMObject> path) {
+    return _fileDialog.setFolder(path);
+  }
+
+  /// Sets the file [name] that is initially shown in the IFileDialog.
+  int setFileName(String name) {
+    return _fileDialog.setFileName(TEXT(name));
+  }
+
+  /// Sets the allowed file type extensions in the IFileOpenDialog.
+  int setFileTypes(int count, Pointer<COMDLG_FILTERSPEC> filters) {
+    return _fileDialog.setFileTypes(count, filters);
+  }
+
+  /// Sets the label of the confirmation button.
+  int setOkButtonLabel(String text) {
+    return _fileDialog.setOkButtonLabel(TEXT(text));
+  }
+
+  /// Returns the HRESULT of getting the IFileDialog's [options](https://pub.dev/documentation/win32/latest/winrt/FILEOPENDIALOGOPTIONS-class.html).
+  /// It also uses the [outOptions] paramater to return the current options.
+  int getOptions(Pointer<Uint32> outOptions) {
+    return _fileDialog.getOptions(outOptions);
+  }
+
+  /// Sets the [options](https://pub.dev/documentation/win32/latest/winrt/FILEOPENDIALOGOPTIONS-class.html)
+  /// given into an IFileDialog.
+  int setOptions(int options) {
+    return _fileDialog.setOptions(options);
+  }
+
+  /// Shows an IFileDialog using the given parent.
+  int show(int parent) {
+    return _fileDialog.show(parent);
+  }
+
+  /// Return results from an IFileDialog. This should be used when selecting
+  /// single items.
+  int getResult(Pointer<Pointer<COMObject>> outItem) {
+    return _fileDialog.getResult(outItem);
+  }
+
+  /// Return results from an IFileOpenDialog. This should be used when selecting
+  /// single items. This function will fail if the IFileDialog* provided to the
+  /// constructor was not an IFileOpenDialog instance, returning an E_NOINTERFACE
+  /// error.
+  int getResults(Pointer<Pointer<COMObject>> outItems) {
+    IFileOpenDialog? fileOpenDialog;
+    try {
+      fileOpenDialog = _iFileOpenDialogFactory.from(_fileDialog);
+      final int result = fileOpenDialog.getResults(outItems);
+      return result;
```

Comment:
```suggestion
      return fileOpenDialog.getResults(outItems);
```

also, wouldn't this return avoid calling the free and release? How is the call stack here called?

Reply:
As this function is returning the operation result and the actual `results` within `outItems`, the dialog should be discarded immediately. 

If we modify the function for returning the new dialog, we may apply what you suggested.